docs/rfcs/0002/readme.md

```diff
@@ -37,21 +34,58 @@ The proposal below might take a couple of read-throughs; I've also added a concr
-4. The USER who desires to use the application works with the AUTHOR and the related NNF Container Profile to understand the storage requirements.
-5. The USER submits a WLM job with the #DW container fields populated
-6. WLM runs the job and drives the job through the following stages...
-    1. Proposal: RABBIT validates the #DW container directive by comparing the supplied values to what is listed in the NNF Container Profile. If the USER fails to meet the requirements, the job fails. 
-    2. Pre-run: RABBIT software will:
-        1. create a config map reflecting the storage requirements and any runtime parameters; this is provided to the container at the volume mount named "nnf-config", if specified.
-        2. duplicate the pod template specification from the Container Profile and patches the necessary Volumes and the config map. The spec is used as the basis for starting the necessary pods and containers.
-    3. The containerized application executes. The expected mounts are available per the requirements and celebration occurs.
-
-Example
--------
-
-Say I authored a simple application, `foo`, that requires Rabbit local GFS2 storage and a persistent Lustre storage volume. As the author, my program is coded to expect the GFS2 volume is mounted at `/foo/local` and the Lustre volume is mounted at `/foo/persistent`. In this case, the storages are not optional, so they are defined as such in the NNF Container Profile.
+    3. the NNF storage requirements described above.
+3. The ADMINISTRATOR creates a corresponding _NNF Container Profile_ Kubernetes custom resource with the necessary NNF storage requirements and pod specification as described by the AUTHOR
+4. The USER who desires to use the application works with the AUTHOR and the related NNF Container Profile to understand the storage requirements
+5. The USER submits a WLM job with the #DW container directive variables populated
+6. WLM runs the workflow and drives it through the following stages...
+    1. `Proposal`: RABBIT validates the #DW container directive by comparing the supplied values to those listed in the NNF Container Profile. If the workflow fails to meet the requirements, the job fails
+    2. `PreRun`: RABBIT software:
+        1. creates a config map reflecting the storage requirements and any runtime parameters; this is provided to the container at the volume mount named `nnf-config`, if specified
+        2. duplicates the pod template specification from the Container Profile and patches the necessary Volumes and the config map. The spec is used as the basis for starting the necessary pods and containers.
+    3. The containerized application(s) executes. The expected mounts are available per the requirements and celebration occurs. The pods continue to run until:
+       1. the pod completes successfully (any failed pods will be restarted)
```

Comment:
Doesn't "any failed pods will be restarted" depend on the restart policy?

Reply:
It's a bit confusing b/c we're not truly restarting the pod - we're creating a _new_ pod to try again. For the purposes of this RFC, I'm calling that a restart. 

In order to keep the failed pods around and allow for debugging/log inspection, we need to use `restartPolicy: Never`. I don't think we should even allow `restartPolicy: OnFailure`.

Reply:
I think I need a summary of our slack conversations
- If I write an application that can successful restart after encountering a crash, is that`OnFailure`?
- If I write an application that does not support restarts, is that `Never`?

Reply:
- `OnFailure`: The pods will actually restart. This means that the pod restart count will increase. k8s will not fire up a new pod, but it will restart the same pod. You'll lose the log from any failures because the pod restarts.
- `Never`: Instead of a restart, you'll get a brand new pod. The logs from the old pods stick around.

So for both of your cases, you'd want `Never'.

For case #1, you set `Job.Spec.backoffLimit=N` where `N` is however many attempts you'd like to allow. This defaults to 6 in k8s.
For case #2, you set Job.Spec.backoffLimit=0`

Remember that by definition, jobs are in the business of firing up pods until a certain number of completions are hit, so it's always going to try retry. The `restartPolicy` is not telling the Job to retry, but rather _how_ to perform the retry. It's asking if it should try to restart the pod or fire up a new pod. We want new pods so that we can keep the logging history.

> A Job creates one or more Pods and will continue to retry execution of the Pods until a specified number of them successfully terminate. As pods successfully complete, the Job tracks the successful completions. When a specified number of successful completions is reached, the task (ie, Job) is complete. Deleting a Job will clean up the Pods it created. Suspending a Job will delete its active Pods until the Job is resumed again.

> A simple case is to create one Job object in order to reliably run one Pod to completion. The Job object will start a new Pod if the first Pod fails or is deleted (for example due to a node hardware failure or a node reboot).

More detail here:
https://kubernetes.io/docs/concepts/workloads/controllers/job/#handling-pod-and-container-failures

Reply:
OK - that's very helpful. Thanks. Some form of this _can_ make it into the document, we've done things like collapsed details in the initial setup docs

I'm trying to shelter the customer from the k8s implementation. Our version of "restartPolicy" that meets their requirements doesn't need to match how it's implemented. But perhaps we introduce new names to avoid any confusion. 

And if I recall from our conversations, we can not control the back-off time in between restarts? We probably need to communicate this to LLNL as a limitation and ensure they are OK with it. I suggest using a question call-out

> You'll lose the log from any failures because the pod restarts.

Does the --previous flag help here?

Reply:
Also just wanted to point out that when we fire up new pods, they get a different IP. But if we have a headless service and set `hostname/subdomain` on these pods, other rabbit nodes and compute nodes can get the IP via dns.

If that's an issue, then we'd need to go back to `OnFailure`, but then that means we lose history/logging.

Reply:
> But perhaps we introduce new names to avoid any confusion.

I actually did this at first, but then went back to the k8s implementation names in case they wanted to look this up. But I think you're right. This stuff is confusing, and it'd be better if we made it less confusing by hiding the implementation.

> And if I recall from our conversations, we can not control the back-off time in between restarts? We probably need to communicate this to LLNL as a limitation and ensure they are OK with it. I suggest using a question call-out

That's just for _actual_ restarts when using `restartPolicy: OnFailure`. You are correct in that you cannot control the back-off time. More details: https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy

With `restartPolicy: Never`, we get new pods ASAP - there is no back-off time between retries in this case, so it would not be an issue. And since our Job only targets 1 node, we can guarantee that the new pod will land on the same node. The downside is that the pod gets a new IP (as I pointed out in my previous comment).